src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/GenerateMethodCodeActionResolver.cs

```diff
@@ -0,0 +1,110 @@
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using CSharpSyntaxFactory = Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Razor;
+internal class GenerateMethodCodeActionResolver : IRazorCodeActionResolver
+{
+    public string Action => LanguageServerConstants.CodeActions.GenerateMethod;
+    private readonly DocumentContextFactory _documentContextFactory;
+    private static readonly string s_generatedCode = "0private void 2()\r\n0{\r\n1throw new NotImplementedException();\r\n0}";
+
+    public GenerateMethodCodeActionResolver(DocumentContextFactory documentContextFactory)
+    {
+        _documentContextFactory = documentContextFactory;
+    }
+
+    public async Task<WorkspaceEdit?> ResolveAsync(JObject data, CancellationToken cancellationToken)
+    {
+        if (data is null)
+        {
+            return null;
+        }
+
+        var actionParams = data.ToObject<GenerateMethodCodeActionParams>();
+        if (actionParams is null)
+        {
+            return null;
+        }
+
+        var documentContext = await _documentContextFactory.TryCreateForOpenDocumentAsync(actionParams.Uri, cancellationToken).ConfigureAwait(false);
+        if (documentContext is null)
+        {
+            return null;
+        }
+
+        var code = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var uriPath = FilePathNormalizer.Normalize(actionParams.Uri.GetAbsoluteOrUNCPath());
+        var codeBehindPath = $"{uriPath}.cs";
+        if (!File.Exists(codeBehindPath))
+        {
+            return GenerateMethodInCodeBlock(code, actionParams);
+        }
+
+        var razorCSharpText = (await documentContext.GetCSharpSourceTextAsync(cancellationToken).ConfigureAwait(false)).ToString();
+        var razorNamespace = razorCSharpText[(razorCSharpText.IndexOf("namespace ") + 10)..razorCSharpText.IndexOf("\r\n{")];
+        var mock = CSharpSyntaxFactory.ParseCompilationUnit(File.ReadAllText(codeBehindPath));
+        var @namespace = mock.Members.Where(m => m is NamespaceDeclarationSyntax { } @namespace && @namespace.Name.ToString() == razorNamespace).FirstOrDefault();
+        if (@namespace is null)
+        {
+            // The code behind file is malformed, generate the code in the razor file instead.
+            return GenerateMethodInCodeBlock(code, actionParams);
+        }
+
+        var @class = ((NamespaceDeclarationSyntax)@namespace).Members.Where
+            (m => m is ClassDeclarationSyntax { } @class && uriPath.LastIndexOf(@class.Identifier.Text) != -1 && uriPath[uriPath.LastIndexOf(@class.Identifier.Text)..uriPath.LastIndexOf('.')] == @class.Identifier.Text)
```

Comment:
Is this just `File.GetFilenameWithoutExtension(uriPath) == @class.Identifier.Text`?

Reply:
Ah it would be `Path.GetFileNameWithoutExtension`, but cool, thanks! Didn't know that this method existed.